// @node-lightning/secp256k1/dist/Secp256k1Error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Secp256k1Error = void 0;
var Secp256k1Error;
(function (Secp256k1Error) {
    Secp256k1Error["IMPOSSIBLE_CASE"] = "Impossible case. Please create issue.";
    Secp256k1Error["TWEAK_ADD"] = "The tweak was out of range or the resulted private key is invalid";
    Secp256k1Error["TWEAK_MUL"] = "The tweak was out of range or equal to zero";
    Secp256k1Error["CONTEXT_RANDOMIZE_UNKNOW"] = "Unknow error on context randomization";
    Secp256k1Error["SECKEY_INVALID"] = "Private Key is invalid";
    Secp256k1Error["PUBKEY_PARSE"] = "Public Key could not be parsed";
    Secp256k1Error["PUBKEY_SERIALIZE"] = "Public Key serialization error";
    Secp256k1Error["PUBKEY_COMBINE"] = "The sum of the public keys is not valid";
    Secp256k1Error["SIG_PARSE"] = "Signature could not be parsed";
    Secp256k1Error["SIGN"] = "The nonce generation function failed, or the private key was invalid";
    Secp256k1Error["RECOVER"] = "Public key could not be recover";
    Secp256k1Error["ECDH"] = "Scalar was invalid (zero or overflow)";
})(Secp256k1Error = exports.Secp256k1Error || (exports.Secp256k1Error = {}));
